src/bit-systems/video-texture.ts

```diff
@@ -0,0 +1,173 @@
+        return;
+      }
+    }
+
+    camera.layers.enable(Layers.CAMERA_LAYER_THIRD_PERSON_ONLY);
+
+    const resolution = VideoTextureSource.resolution[eid];
+    camera.aspect = resolution[0] / resolution[1];
+
+    // TODO currently if a video-texture-source tries to render itself it will fail with a warning.
+    // If we want to support this we will need 2 render targets to swap back and forth.
+    const renderTarget = new WebGLRenderTarget(resolution[0], resolution[1], {
+      format: RGBAFormat,
+      minFilter: LinearFilter,
+      magFilter: NearestFilter,
+      encoding: sRGBEncoding
+    });
+
+    // Since we are rendering directly to a texture we need to flip it vertically
+    // See https://github.com/mozilla/hubs/pull/4126#discussion_r610120237
+    renderTarget.texture.matrixAutoUpdate = false;
+    renderTarget.texture.matrix.scale(1, -1);
+    renderTarget.texture.matrix.translate(0, 1);
+
+    sourceDataMap.set(eid, { renderTarget, lastUpdated: 0, camera: camera.eid! });
+  });
+  exitedVideoTextureSourcesQuery(world).forEach(function (eid) {
+    const srcData = sourceDataMap.get(eid);
+    if (srcData) {
+      srcData.renderTarget.dispose();
+      sourceDataMap.delete(eid);
+    }
+  });
+
+  enteredVideoTextureTargetsQuery(world).forEach(function (eid) {
+    const srcData = sourceDataMap.get(VideoTextureTarget.srcNode[eid]);
+    const targetData: TargetData = { originalMap: null, originalEmissiveMap: null };
+    targetDataMap.set(eid, targetData);
+
+    if (!srcData) {
+      console.error("video-texture-target unable to find source");
+      return;
+    }
+
+    const mat = world.eid2mat.get(eid)! as MeshStandardMaterial;
+    if (VideoTextureTarget.flags[eid] & VIDEO_TEXTURE_TARGET_FLAGS.TARGET_BASE_MAP) {
+      targetData.originalMap = mat.map;
+      mat.map = srcData.renderTarget.texture;
+    }
+    if (VideoTextureTarget.flags[eid] & VIDEO_TEXTURE_TARGET_FLAGS.TARGET_EMISSIVE_MAP) {
+      targetData.originalEmissiveMap = mat.emissiveMap;
+      mat.emissiveMap = srcData.renderTarget.texture;
+    }
+  });
+  exitedVideoTextureTargetsQuery(world).forEach(function (eid) {
+    const targetData = targetDataMap.get(VideoTextureTarget.srcNode[eid])!;
+    const mat = world.eid2mat.get(eid)! as MeshStandardMaterial;
+    if (VideoTextureTarget.flags[eid] & VIDEO_TEXTURE_TARGET_FLAGS.TARGET_BASE_MAP) {
+      mat.map = targetData.originalMap;
+    }
+    if (VideoTextureTarget.flags[eid] & VIDEO_TEXTURE_TARGET_FLAGS.TARGET_EMISSIVE_MAP) {
+      mat.emissiveMap = targetData.originalMap;
+    }
+    targetDataMap.delete(eid);
+  });
+
+  videoTextureSourceQuery(world).forEach(function (eid) {
+    const data = sourceDataMap.get(eid);
+    if (data && world.time.elapsed > data.lastUpdated + 1000 / VideoTextureSource.fps[eid]) {
+      updateRenderTarget(world, data.renderTarget, data.camera);
+      data.lastUpdated = world.time.elapsed;
```

Comment:
Why does video-texture.ts set `renderTarget.needsUpdate = false` after `updateRenderTarget` whereas that does not happen here?

Reply:
That was just an additional property I sloppily stuck on `renderTarget` in the camera-tool system. This system does not use that, though the purpose of that in camera-tool system was to only update the render target when the material preview is visible, which we should actually do here too...